cpp/src/reductions/nth_element.cu

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// The translation unit for reduction `min`
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/reduction_functions.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <rmm/device_scalar.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace reduction {
+namespace {
+
+template <typename ElementType>
+struct nth_element_functor {
+  template <typename OutputType>
+  static constexpr bool is_supported_v() {
+    return std::is_convertible<ElementType, OutputType>::value;
+  }
+
+  template <typename OutputType>
+  std::enable_if_t<(is_supported_v<OutputType>() &&
+                    is_fixed_width<OutputType>()),
+                   std::unique_ptr<scalar>>
+  operator()(column_device_view const& dcol,
+             size_type n,
+             rmm::mr::device_memory_resource* mr,
+             cudaStream_t stream) {
+    bool const n_valid = n < dcol.size() && n >= -dcol.size();
+    using ScalarType = cudf::experimental::scalar_type_t<OutputType>;
+    auto result = new ScalarType(OutputType{0}, n_valid, stream, mr);
+    if (n_valid) {
+      n = (n < 0 ? dcol.size() + n : n);
+      //device to device copy
+      thrust::for_each_n(
+          rmm::exec_policy(stream)->on(stream),
+          thrust::make_counting_iterator<size_type>(n),1,
+          [dres = result->data(), dvalid = result->validity_data(),
+           dcol] __device__(auto n) {
+            *dres = static_cast<OutputType>(dcol.element<ElementType>(n));
+            *dvalid = dcol.is_valid(n);
+          });
+    }
+    return std::unique_ptr<scalar>(result);
+  }
+
+  template <typename OutputType>
+  static constexpr bool is_supported_s() {
```

Comment:
maybe rename this to something like `is_string_input_output`, so it's easier to read the code.